File: 86-partition-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode * dummy = new ListNode(-1);
        dummy->next = head;
        ListNode * prev = dummy, * curr = head;
        while (prev->next && prev->next->val < x) prev = prev->next;
        curr = prev;
        while (curr->next) {
            if (curr->next->val < x) {
                ListNode * temp = curr->next;
                curr->next = temp->next;
                temp->next = prev->next;
                prev->next = temp;
                prev = temp;
            } else {
                curr = curr->next;
            }
        }
        return dummy->next;
    }   
};

File: 86-partition-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        dummy1, dummy2 = ListNode(-1), ListNode(-1)
        p1, p2 = dummy1, dummy2
        p = head
        while (p):
            if p.val >= x:
                p2.next = p
                p2 = p2.next
            if p.val < x:
                p1.next = p
                p1 = p1.next
        
            temp = p.next
            p.next = None
            p = temp
        
        p1.next = dummy2.next
        return dummy1.next
